POEM_069.md

```diff
@@ -0,0 +1,167 @@
+        for i_conn in range(self.options['n_out']):
+            residuals['V'] -= inputs['I_out:{}'.format(i_conn)]
+```
+
+In this Node component the residual is the net current at the node.
+Yet the key for the residual is `V` because it is set using the output variable names. 
+It would be more intuitive to name the residual `I_net`. 
+
+In multiple dimensions, this connection between output/state name and residual name becomes even more confusing.
+
+```python
+
+class Confusing(om.ImplicitComponent):
+    def setup(self):
+        self.add_output("x")
+        self.add_output("y")
+
+        self.declare_partials("x", ["x", "y"])
+        self.declare_partials("y", ["x"])
+
+    def apply_nonlinear(self, inputs, outputs, residuals):
+        x = outputs["x"]
+        y = outputs["y"]
+        residuals["x"] = y * x - x ** 2
+        residuals["y"] = x + 3.1416
+
+    def linearize(self, inputs, outputs, partials):
+        x = outputs["x"]
+        y = outputs["y"]
+        partials["x", "x"] = y - 2 * x
+        partials["x", "y"] = x
+        partials["y", "x"] = 1    
+```
+
+In this example, the fact that one residual is accessed with `"x"` and the other with `"y"` is arbitrary.
+There is no apparent relationship between the output names and the residuals.
+For example, `residuals["y"]` has no relation to `y`.
+Declaring and implementing the derivatives can be even more confusing because the keys don't imply a relationship with the residuals.
+A new user would look at `partials["x", "x"]` and wonder why it isn't equal to 1 (speaking from experience...).
+
+
+## Proposed solutions
+
+We propose adding the implicit component method `add_residuals` to allow the user to explicitly declare the residuals.
+This would decouple the residual names and shapes from the output names and shapes, making it more intuitive for new users.
+Furthermore, this API modification is consistent with the way that inputs and outputs are declared and used by the methods.
+
+
+```python
+class Node(om.ImplicitComponent):
+    """Computes voltage residual across a node based on incoming and outgoing current."""
+
+    def initialize(self):
+        self.options.declare('n_in', default=1, types=int, desc='number of connections with + assumed in')
+        self.options.declare('n_out', default=1, types=int, desc='number of current connections + assumed out')
+
+    def setup(self):
+        self.add_output('V', val=5., units='V')
+
+        for i in range(self.options['n_in']):
+            i_name = 'I_in:{}'.format(i)
+            self.add_input(i_name, units='A')
+            self.declare_partials('V', i_name, val=1)
+
+        for i in range(self.options['n_out']):
+            i_name = 'I_out:{}'.format(i)
+            self.add_input(i_name, units='A')
+            self.declare_partials('V', i_name, val=-1)
+        
+        self.add_residual('I_net', val=0., desc='net current flowing through the node')
```

Comment:
In the description of `add_residual` that comes later, it says the args are (name, shape), but here it also takes 'val' and 'desc'.  The POEM should include the full list of arguments that can be passed to `add_residual`.  Also, the existing `add_output` method supports `res_ref` for residual scaling and `res_units`.  Will the `add_residual` method have corresponding `ref` and `units` args?

Reply:
Agreed. Would `val` get used anywhere if we did specify it? As far as I know, it wouldn't be used by a monolithic solver, but maybe for another case?

Reply:
I can't think of any cases where `val` would actually be used.  We probably just shouldn't allow `val` as an arg.

Reply:
Agreed. Would `val` be used by the solver somehow if specified? From my understanding of the monolithic solver it would not, but I'm not sure if it may be different for other cases.

Reply:
Ok, sounds good. We'll make the change to the POEM. (sorry about the duplicate comment, my computer was struggling)